### ISSUE TITLE
Better IAM check and warnings when creating HTTP functions

### ISSUE THREAD

**Issue by mbleigh** [firebase/firebase-tools#2086]
Due to the recent change in default IAM policy for HTTP Cloud Functions, it is now the case that to make a public HTTP function the `cloudfunctions.functions.setIamPolicy` permission is required. This permission is included in `owner`, but not `editor`.

We should perform a specialized IAM check when creating an HTTP function so we can properly warn users:

1. If any of the actions to be taken by a deploy will be the creation of an HTTP function, then...
2. Run a `testIamPermissions` for the user at the project level to see if they have `cloudfunctions.functions.setIamPolicy`.
3. If they do, allow the deploy to proceed.
4. If they don't, error out with a clear error message (see below):

```
Creating HTTP functions requires permission cloudfunctions.functions.setIamPolicy which
was not detected for these credentials. This permission can be granted by 'owner' and
'cloudfunctions.admin' roles in the Firebase or Google Cloud consoles.
```

I'm on the fence about whether to make this an error or a warning. It's far from ideal in either case, honestly. I'm going to follow up separately to see if we can address this more cleanly.

**Comment by pdemilly** [comment]
I have a callable method for which I have set allUsers permissions for Cloud functions invoker but still get a CORS error

Access to fetch at 'https://xxx.com/xxx-3da7f/us-central1/reportFn-generateReportUrl' from origin 'http://localhost:4200' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

Debugging the output, I can see that it is not interpreted as a cloud function but at request to my hosted website. So something is changed in the way new URL functions are rewritten on the backend

**Comment by mbleigh** [comment]
Improvements to error messages have been released in `v8.0.2`.

**Comment by Shajeel-Afzal** [comment]
I'm facing this problem. Is there any document that instructs how to assign this permission in the Google Cloud Console?
